### PR TITLE
Canvas: New basic elements

### LINKED ISSUE #83270: Canvas: Add most impactful element types that users expect in flow charts
1. Triangle
2. Cloud
3. Parallelogram

### PR DESCRIPTION
**Pull request by Develer**
**What is this feature?**

This PR implements new canvas elements like triangles, clouds, and parallelograms.

**Which issue(s) does this PR fix?**:

Fixes #83270 #74903

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

https://github.com/grafana/grafana/assets/1418924/8d1db29b-292a-4951-a5a0-bc74824f1c2d

Added new shapes to existing canvas gdev dashboard for testing data links

https://github.com/grafana/grafana/assets/22381771/3041c7e0-b060-4201-a20e-04d01e4479ed

### CODE REVIEW

**Review by nmarrs — COMMENTED** [review]
Great job on these @Develer!


https://github.com/grafana/grafana/assets/22381771/f3421c10-2231-480d-9c85-2edafcc7990d



I think we need to figure out the question of how we handle these elements styling properties (such as background image / color / border)

With rectangle it is easy as the element fills the entire div making it very straightforward. However with ellipse and now these three added elements it is not clear that the styling should apply to the background of the element vs the element itself

IMO these settings should probably apply to the element themselves vs the background. How we address this might be a little tricky and is dependent on [the WIP work here refactoring ellipse's approach to styling options](https://github.com/grafana/grafana/pull/83714) 

What I'm not sure about is providing ability to modify both element style and background styles and whether or not we need to update moveable's selection rendering (this might be very tricky, and maybe not necessary given draw.io does a box selection approach regardless of the element)

**TLDR; I think we really need to figure out the current WIP of ellipse styling refactor, get to a "good enough" resolution for the time being to unblock this work**

cc @lukasztyrala - interested in your thoughts here

File: public/app/features/canvas/elements/parallelogram.tsx [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import { Align, TextConfig, TextData, VAlign } from '../types';
+
+class Parallelogram extends PureComponent<CanvasElementProps<TextConfig, TextData>> {
+  render() {
+    const { data } = this.props;
+    const styles = getStyles(config.theme2, data);
+
+    return (
+      <div className={styles.container}>
+        <svg
```
Comment (nmarrs) [inline public/app/features/canvas/elements/parallelogram.tsx:21]: Any reason why parrellogram / cloud have default borders set already (the rest of elements in canvas don't do this 🤔)

Screenshot 2024-03-13 at 9 30 40 AM

File: public/app/features/canvas/elements/triangle.tsx [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+const getStyles = (theme: GrafanaTheme2, data: TextData | undefined) => ({
+  container: css({
+    height: '100%',
+    width: '100%',
+  }),
+  text: css({
+    position: 'absolute',
+    top: data?.valign === VAlign.Middle ? '50%' : data?.valign === VAlign.Top ? '10%' : '90%',
+    left: data?.align === Align.Center ? '50%' : data?.align === Align.Left ? '10%' : '90%',
+    transform: `translate(${data?.align === Align.Center ? '-50%' : data?.align === Align.Left ? '10%' : '-90%'}, ${data?.valign === VAlign.Middle ? '-50%' : data?.valign === VAlign.Top ? '10%' : '-90%'})`,
```
[the remark this answers is not in the document]
Comment (Develer) [reply public/app/features/canvas/elements/triangle.tsx:160]: there is deffernet values and numbers, I don't see how to optimize it better, maybe only first two lines, but not sure if it worth it

File: public/app/features/canvas/elements/triangle.tsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import { Align, TextConfig, TextData, VAlign } from '../types';
+
+class Triangle extends PureComponent<CanvasElementProps<TextConfig, TextData>> {
+  render() {
+    const { data } = this.props;
+    const styles = getStyles(config.theme2, data);
+
+    return (
+      <div className={styles.container}>
+        <svg
```
[the remark this answers is not in the document]
Comment (Develer) [reply public/app/features/canvas/elements/triangle.tsx:21]: I see, I'll adjust shape a bit.

**Review by baldm0mma — APPROVED** [review]
LGTM!

**Comment by Develer** [comment]
> From user's perspective, it will be expected that:
> 
> 1. The element's border option controls the border of the shape, not the container.
> 2. The element's background option controls the color of the element, not the container.
> 3. The background coloring based on Fields applies to the background of the element, not the container.
> 
> This is something we should address in future iterations. Apart from that, LGTM.

Yes, that's what I already done in one of the last commits. At least ellipse, triangle, cloud, and parallelogram shapes have their own (not a container's) border and background.

File: public/app/features/canvas/elements/cloud.tsx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  },
+};
+
+const getStyles = (theme: GrafanaTheme2, data: CanvasElementData | undefined) => ({
+  container: css({
+    height: '100%',
+    width: '100%',
+  }),
+  text: css({
+    position: 'absolute',
```
Comment (codeincarnate) [inline public/app/features/canvas/elements/cloud.tsx:163]: Computing these booleans outside of the properties would make this more readable. Especially since there's only a handful of values

File: public/app/features/canvas/elements/ellipse.tsx [hunk 5]
```diff
@@ -194,3 +152,23 @@ export const ellipseItem: CanvasElementItem<EllipseConfig, EllipseData> = {
...
 };
+
+const getStyles = (theme: GrafanaTheme2, data: CanvasElementData | undefined) => ({
+  container: css({
+    height: '100%',
+    width: '100%',
+  }),
+  text: css({
+    position: 'absolute',
+    top: data?.valign === VAlign.Middle ? '50%' : data?.valign === VAlign.Top ? '10%' : '90%',
```
Comment (codeincarnate) [inline public/app/features/canvas/elements/ellipse.tsx:163]: Probably shouldn't repeat the logic for this either 😄

**Review by codeincarnate — APPROVED** [review]
Looks pretty good! Did notice an issue with the cloud being cut off with larger border sizes. Also text layout on the triangle can get a little wonky but those can both be left for follow-ups I think 😄 

Just an example of what I mean with the cloud:

Screenshot 2024-03-19 at 11 11 46 AM

Also, is the next step to add support for custom SVG elements 🤔 😆

File: public/app/features/canvas/elements/ellipse.tsx (hunk 5 above)
Comment (Develer) [reply public/app/features/canvas/elements/ellipse.tsx:185]: agree, let me move it outside

**Comment by Develer** [comment]
> Looks pretty good! Did notice an issue with the cloud being cut off with larger border sizes. Also text layout on the triangle can get a little wonky but those can both be left for follow-ups I think 😄
> 
> Just an example of what I mean with the cloud:
> 
> Screenshot 2024-03-19 at 11 11 46 AM
> Also, is the next step to add support for custom SVG elements 🤔 😆

yes, I'm aware of it, I didn't find a quick solution, so definitely for a follow-up task 🙂

File: public/app/features/canvas/elements/cloud.tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+  defaultTextColor,
+} from '../element';
+import { Align, CanvasElementConfig, CanvasElementData, VAlign } from '../types';
+
+const Cloud = (props: CanvasElementProps<CanvasElementConfig, CanvasElementData>) => {
+  const { data } = props;
+  const styles = getStyles(config.theme2, data);
+
+  // uuid needed to avoid id conflicts when multiple elements are rendered
+  const [uniqueId] = useState(() => uuidv4());
```
[the remark this answers is not in the document]
Comment (nmarrs) [reply public/app/features/canvas/elements/cloud.tsx]: Addressed this as part of https://github.com/grafana/grafana/pull/84205/commits/1748afe2bee0453b4cbb7e5c8f3a6d2a1c9c979d

File: public/app/features/canvas/runtime/element.tsx [hunk 7]
```diff
@@ -191,8 +193,13 @@ export class ElementState implements LayerElement {
         this.div.style[key as any] = (this.sizeStyle as any)[key];
       }
 
-      for (const key in this.dataStyle) {
-        this.div.style[key as any] = (this.dataStyle as any)[key];
+      const elementType = this.options.type;
+      // SVG elements have their own styles
+      // TODO: This is a hack, we should have a better way to handle this
```
Comment (nmarrs) [inline public/app/features/canvas/runtime/element.tsx:198]: This will lead to some glitches / bugs, such as when changing element types from rectangle to one of the svg elements

We should create a follow-up github issue to address this bug ideally as part bug bash / feature hardening phase

https://github.com/grafana/grafana/assets/22381771/c779e3ed-131a-4fdc-a865-26272c0db51b

File: public/app/features/canvas/types.ts [hunk 8]
```diff
@@ -48,16 +48,20 @@ export interface TextConfig {
...
   backgroundColor?: ColorDimensionConfig;
+  backgroundImage?: ResourceDimensionConfig;
+  backgroundSize?: BackgroundImageSize;
   borderColor?: ColorDimensionConfig;
-  width?: number;
+  borderWidth?: number;
 }
 
-export interface EllipseData extends TextData {
+export interface CanvasElementData extends TextData {
```
Comment (nmarrs) [inline public/app/features/canvas/types.ts:59]: We shouldn't need either of these anymore as `rectangle`, `ellipse`, `triangle`, `paralleogram`, and `cloud` should all be based on the `TextConfig` 😬

**Review by nmarrs — APPROVED** [review]
I will work on adding the migration code accounting for ellipse options changes and then I think this is in a "good enough" state to be merged, hopefully we can revisit the known bugs / issues as part of bug cleanup for actual G11 release build 😬

File: public/app/features/canvas/runtime/element.tsx (hunk 7 above)
Comment (nmarrs) [reply public/app/features/canvas/runtime/element.tsx:198]: https://github.com/grafana/grafana/issues/84869

File: public/app/features/canvas/types.ts (hunk 8 above)
Comment (nmarrs) [reply public/app/features/canvas/types.ts:59]: we can disregard this for now, but we probably should figure out maybe a different way of passing the base elements config (i.e. borderWidth, background, etc) down instead of overlaying it with the element specific config / data

**Comment by nmarrs** [comment]
I tried to address an issue I noticed where transparency background color is displayed as all black / is broken for the svg elements

I created [this github issue ](https://github.com/grafana/grafana/issues/84871)to follow up on this

File: public/app/plugins/panel/canvas/migrations.ts [hunk 9]
```diff
@@ -18,5 +18,25 @@ export const canvasMigrationHandler = (panel: PanelModel): Partial<Options> => {
...
+        if (element.type === 'ellipse') {
+          // Take existing ellipse specific background and border config and apply it to the element's general background and border config
+          element.background = element.config.backgroundColor;
+          element.border.color = element.config.borderColor;
+          element.border.width = element.config.width;
+
+          // Remove the ellipse specific background and border config
+          delete element.config.backgroundColor;
+          delete element.config.borderColor;
+          delete element.config.width;
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/migrations.ts:35]: Example of ellipse migration. To test locally you will need to create an ellipse on main / hard code old config and then set the pluginVersion in the dashboard json to `11.0.1` to force the migration code to run

https://github.com/grafana/grafana/assets/22381771/d9872e17-9079-496b-939c-bb7fe716bdfb
Screenshot 2024-03-20 at 6 02 57 PM

Before
before


After (No migration code)
no migration code


After (With migration code)

after migration

**Comment by nmarrs** [comment]
I'm going to defer to @Develer to merge this / run through and test things to make sure my changes didn't break anything 😬 I think we have a fair bit to come back to and address as part of polishing this up for main release 😅

File: public/app/plugins/panel/canvas/migrations.ts (hunk 9 above)
Comment (Develer) [reply public/app/plugins/panel/canvas/migrations.ts:35]: Thanks for adding migration @nmarrs ! Looks and works fine for me, didn't notice any odd behavior.
